File: golang/221-Maximum_Square.go
package golang

func maximalSquare(matrix [][]byte) int {
    if len(matrix) == 0 || len(matrix[0]) == 0 {
        return 0
    }
    m, n := len(matrix), len(matrix[0])
    dp := make([][]int, m)
    for i := range dp {
        dp[i] = make([]int, n)
    }
    var maxL int
    for i := 0; i < m; i ++ {
        for j := 0; j < n; j ++ {
            if matrix[i][j] == '1' {
                if i == 0 || j == 0{
                    dp[i][j] = 1
                } else {
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])+1
                }
                if maxL < dp[i][j] {
                    maxL = dp[i][j]
                }
            }
        }
    }
    return maxL * maxL
}

func min(a ...int) int {
    min := math.MaxInt32
    for i := 0; i < len(a); i ++ {
        if a[i] < min {
            min = a[i]
        }
    }
    return min
}

File: Top_100_Liked/221-Maximum_Square.cpp
#include <bits/stdc++.h>
using namespace std;


class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m = matrix.size();
        if (m == 0) return 0;
        int n = matrix[0].size();
        int dp[m+1][n+1] = {0};
        int maxL = 0;
        for (int i = 1; i <= m; i ++ ) {
            for (int j = 1; j <= n; j ++ ) {
                if (matrix[i-1][j-1] == '1') {
                    dp[i][j] = min(dp[i-1][j], min(dp[i-1][j-1], dp[i][j-1])) + 1;
                    maxL = max(maxL, dp[i][j]);
                }
            }
        }
        return maxL * maxL;
    }
};
/* 
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if (matrix.size() == 0) return 0;
        int row = matrix.size();
        int col = matrix[0].size();
        int res = 0;
        for (int i = 0; i < row; i ++ ) {
            for (int j = 0; j < col; j ++ ) {
                int x = i, y = j;
                int xMax = 0, yMax = 0;
                while (y < col) {
                    if (matrix[i][y] == 1) {
                        y ++;
                    } else break;
                }
                if (y - j > 0) {
                    
                }
            }
        }
    }
};
*/